File: auth/mailer.js
const nodemailer = require("nodemailer");

/**
 * Validates SMTP configuration at boot time.
 * Fails fast if SMTP is in a broken partial state.
 */
function validateSmtpConfig() {
  const host = process.env.SMTP_HOST;
  const port = process.env.SMTP_PORT;
  const user = process.env.SMTP_USER;
  const pass = process.env.SMTP_PASS;
  const from = process.env.SMTP_FROM;

  const vars = { SMTP_HOST: host, SMTP_PORT: port, SMTP_USER: user, SMTP_PASS: pass, SMTP_FROM: from };
  const definedCount = Object.values(vars).filter(v => v !== undefined && v !== "").length;

  if (definedCount > 0 && definedCount < 5) {
    const missing = Object.entries(vars).filter(([, v]) => !v).map(([k]) => k);
    throw new Error(
      `FATAL: Incomplete SMTP configuration detected. Missing variables: ${missing.join(", ")}. ` +
      `Either provide all 5 SMTP variables (SMTP_HOST, SMTP_PORT, SMTP_USER, SMTP_PASS, SMTP_FROM) or leave them unset for dev mode.`
    );
  }

  const isConfigured = definedCount === 5;
  return { isConfigured, host, port: parseInt(port, 10) || 587, user, pass, from };
}

const smtpConfig = validateSmtpConfig();
let transporter = null;

if (smtpConfig.isConfigured) {
  transporter = nodemailer.createTransport({
    host: smtpConfig.host,
    port: smtpConfig.port,
    secure: smtpConfig.port === 465,
    auth: {
      user: smtpConfig.user,
      pass: smtpConfig.pass ? smtpConfig.pass.replace(/\s+/g, "") : ""
    }
  });
}

/**
 * Sends a one-time passcode (OTP) email to the target recipient.
 * If SMTP is not configured, logs the OTP to console for frictionless local development.
 *
 * @param {string} to - Recipient email address
 * @param {string} otp - 6-digit one-time password
 * @param {number} expiryMinutes - OTP lifetime in minutes
 * @returns {Promise<{ delivered: boolean, devMode?: boolean, preview?: string }>}
 */
async function sendOtpEmail(to, otp, expiryMinutes = 10) {
  if (!to || typeof to !== "string") {
    throw new Error("Recipient email is required.");
  }

  const subject = "SIH Investigation Portal - Your Login Verification Code";
  const textBody = `Your single-use verification code for the SIH Criminal Investigation Network Platform is: ${otp}\n\nThis code expires in ${expiryMinutes} minutes.\n\nIf you did not attempt this action, notify your system administrator immediately.`;
  const htmlBody = `
    <div style="font-family: Arial, sans-serif; max-width: 500px; margin: 0 auto; padding: 20px; border: 1px solid #e0e0e0; border-radius: 8px;">
      <h2 style="color: #1a365d; margin-top: 0;">SIH Investigation Platform</h2>
      <p style="color: #4a5568; font-size: 14px;">Use the verification code below to complete your authentication.</p>
      <div style="background-color: #edf2f7; padding: 15px; border-radius: 6px; text-align: center; margin: 20px 0;">
        <span style="font-size: 28px; font-weight: bold; letter-spacing: 6px; color: #2b6cb0;">${otp}</span>
      </div>
      <p style="color: #718096; font-size: 12px;">This one-time passcode will expire in <strong>${expiryMinutes} minutes</strong>.</p>
      <hr style="border: 0; border-top: 1px solid #e2e8f0; margin: 20px 0;" />
      <p style="color: #a0aec0; font-size: 11px;">If you did not request this code, please ignore this email or contact the forensic admin team immediately.</p>
    </div>
  `;

  if (!transporter) {
    // Dev-mode fallback: intentional console log
    console.log(`[DEV MAIL] OTP for ${to.trim().toLowerCase()}: ${otp} (valid for ${expiryMinutes}m)`);
    return { delivered: false, devMode: true, preview: otp };
  }

  try {
    const info = await transporter.sendMail({
      from: smtpConfig.from,
      to: to.trim().toLowerCase(),
      subject,
      text: textBody,
      html: htmlBody
    });
    return { delivered: true, messageId: info.messageId };
  } catch (err) {
    console.error(`[SMTP ERROR] Failed to send email to ${to}:`, err.message);
    const smtpError = new Error(`SMTP Dispatch Failed: ${err.message}`);
    smtpError.statusCode = 502;
    throw smtpError;
  }
}

module.exports = {
  sendOtpEmail,
  validateSmtpConfig
};

File: backend/app/security/mailer.py
import os
import smtplib
from pathlib import Path
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import Optional, Dict, Any
from dotenv import load_dotenv

# Load root .env if not already loaded
root_env = Path(__file__).resolve().parent.parent.parent.parent / ".env"
if root_env.exists():
    load_dotenv(dotenv_path=root_env)
else:
    load_dotenv()

def get_smtp_config() -> Dict[str, Any]:
    host = os.getenv("SMTP_HOST")
    port = os.getenv("SMTP_PORT")
    user = os.getenv("SMTP_USER")
    pass_ = os.getenv("SMTP_PASS")
    from_ = os.getenv("SMTP_FROM")

    vars_dict = {
        "SMTP_HOST": host,
        "SMTP_PORT": port,
        "SMTP_USER": user,
        "SMTP_PASS": pass_,
        "SMTP_FROM": from_
    }
    defined_count = sum(1 for v in vars_dict.values() if v is not None and v != "")

    is_configured = defined_count == 5
    port_int = int(port) if port and port.isdigit() else 587
    clean_pass = pass_.replace(" ", "") if pass_ else ""
    return {
        "is_configured": is_configured,
        "host": host,
        "port": port_int,
        "user": user,
        "pass": clean_pass,
        "from": from_
    }

def send_otp_email(to: str, otp: str, expiry_minutes: int = 10) -> Dict[str, Any]:
    if not to or not isinstance(to, str):
        raise ValueError("Recipient email is required.")

    clean_to = to.strip().lower()
    subject = "SIH Investigation Portal - Your Login Verification Code"
    text_body = (
        f"Your single-use verification code for the SIH Criminal Investigation Network Platform is: {otp}\n\n"
        f"This code expires in {expiry_minutes} minutes.\n\n"
        f"If you did not attempt this action, notify your system administrator immediately."
    )
    html_body = f"""
    <div style="font-family: Arial, sans-serif; max-width: 500px; margin: 0 auto; padding: 20px; border: 1px solid #e0e0e0; border-radius: 8px;">
      <h2 style="color: #1a365d; margin-top: 0;">SIH Investigation Platform</h2>
      <p style="color: #4a5568; font-size: 14px;">Use the verification code below to complete your authentication.</p>
      <div style="background-color: #edf2f7; padding: 15px; border-radius: 6px; text-align: center; margin: 20px 0;">
        <span style="font-size: 28px; font-weight: bold; letter-spacing: 6px; color: #2b6cb0;">{otp}</span>
      </div>
      <p style="color: #718096; font-size: 12px;">This one-time passcode will expire in <strong>{expiry_minutes} minutes</strong>.</p>
      <hr style="border: 0; border-top: 1px solid #e2e8f0; margin: 20px 0;" />
      <p style="color: #a0aec0; font-size: 11px;">If you did not request this code, please ignore this email or contact the forensic admin team immediately.</p>
    </div>
    """

    config = get_smtp_config()

    if not config["is_configured"]:
        print(f"\n[DEV MAIL] OTP for {clean_to}: {otp} (valid for {expiry_minutes}m)\n")
        return {"delivered": False, "dev_mode": True, "preview": otp}

    try:
        msg = MIMEMultipart("alternative")
        msg["Subject"] = subject
        msg["From"] = config["from"]
        msg["To"] = clean_to

        part1 = MIMEText(text_body, "plain")
        part2 = MIMEText(html_body, "html")
        msg.attach(part1)
        msg.attach(part2)

        if config["port"] == 465:
            with smtplib.SMTP_SSL(config["host"], config["port"], timeout=10) as server:
                server.login(config["user"], config["pass"])
                server.sendmail(config["from"], [clean_to], msg.as_string())
        else:
            with smtplib.SMTP(config["host"], config["port"], timeout=10) as server:
                server.starttls()
                server.login(config["user"], config["pass"])
                server.sendmail(config["from"], [clean_to], msg.as_string())

        return {"delivered": True, "recipient": clean_to}
    except Exception as e:
        print(f"[SMTP ERROR] Failed to send email to {clean_to}: {e}")
        return {"delivered": False, "error": str(e), "dev_mode": True, "preview": otp}
